lib/loader/loader.c

```diff
@@ -44,8 +57,30 @@ static void strcpy_s(char* dest, size_t destsz, const char* src)
 }
 #endif
 
+#define PATH_MAX_LENGTH 4096
+
 static const char* last_error_msg = NULL;
 
+#define LAST_ERROR_MSG_BUFFER_SIZE 511
+
+// Buffer for formatted error messages.
+// It has one null byte extra to avoid buffer read overflow during concurrent access.
+static char last_error_msg_buffer[LAST_ERROR_MSG_BUFFER_SIZE + 1];
+
+ATTR_FORMAT(printf, 2, 3)
+static enum evmc_loader_error_code set_error(enum evmc_loader_error_code error_code,
+                                             const char* format,
+                                             ...)
+{
+    va_list args;
+    va_start(args, format);
+    if (vsnprintf(last_error_msg_buffer, LAST_ERROR_MSG_BUFFER_SIZE, format, args) <
+        LAST_ERROR_MSG_BUFFER_SIZE)
+        last_error_msg = last_error_msg_buffer;
+    va_end(args);
+    return error_code;
```

Comment:
what's the point of passing `error_code` here and doing nothing with it, just returning back?

Reply:
It's only to make usage of it nicer
```
ec = set_error(EVMC_LOADER_INVALID_ARGUMENT, "invalid argument: file name cannot be empty");
```
Note it's an internal function.

Reply:
Additional step for the code reader to figure out what's happening...